Mathlib/Geometry/Manifold/IsImmersionEmbedding.lean

```diff
@@ -0,0 +1,255 @@
+    (hcodChart : codChart ∈ IsManifold.maximalAtlas I' n M')
+    (hwrittenInExtend : EqOn ((codChart.extend I') ∘ f ∘ (domChart.extend I).symm) (equiv ∘ (·, 0))
+      (domChart.extend I).target) : IsImmersionAt F I I' n f x :=
+  LiftSourceTargetPropertyAt.mk_of_continuousAt hf isLocalSourceTargetProperty_immersionAtProp
+    _ _ hx hfx hdomChart hcodChart ⟨equiv, hwrittenInExtend⟩
+
+/-- A choice of chart on the domain `M` of an immersion `f` at `x`:
+w.r.t. this chart and the data `h.codChart` and `h.equiv`,
+`f` will look like an inclusion `u ↦ (u, 0)` in these extended charts.
+The particular chart is arbitrary, but this choice matches the witnesses given by
+`h.codChart` and `h.codChart`. -/
+noncomputable def domChart (h : IsImmersionAt F I I' n f x) : PartialHomeomorph M H :=
+  LiftSourceTargetPropertyAt.domChart h
+
+/-- A choice of chart on the co-domain `N` of an immersion `f` at `x`:
+w.r.t. this chart and the data `h.domChart` and `h.equiv`,
+`f` will look like an inclusion `u ↦ (u, 0)` in these extended charts.
+The particular chart is arbitrary, but this choice matches the witnesses given by
+`h.equiv` and `h.domChart`. -/
+noncomputable def codChart (h : IsImmersionAt F I I' n f x) : PartialHomeomorph M' H' :=
+  LiftSourceTargetPropertyAt.codChart h
+
+lemma mem_domChart_source (h : IsImmersionAt F I I' n f x) : x ∈ h.domChart.source :=
+  LiftSourceTargetPropertyAt.mem_domChart_source h
+
+lemma mem_codChart_source (h : IsImmersionAt F I I' n f x) : f x ∈ h.codChart.source :=
+  LiftSourceTargetPropertyAt.mem_codChart_source h
+
+lemma domChart_mem_maximalAtlas (h : IsImmersionAt F I I' n f x) :
+    h.domChart ∈ IsManifold.maximalAtlas I n M :=
+  LiftSourceTargetPropertyAt.domChart_mem_maximalAtlas h
+
+lemma codChart_mem_maximalAtlas (h : IsImmersionAt F I I' n f x) :
+    h.codChart ∈ IsManifold.maximalAtlas I' n M' :=
+  LiftSourceTargetPropertyAt.codChart_mem_maximalAtlas h
+
+lemma map_source_subset_source (h : IsImmersionAt F I I' n f x) :
+    f '' h.domChart.source ⊆ h.codChart.source :=
+  LiftSourceTargetPropertyAt.map_source_subset_source h
+
+/-- A linear equivalence `E × F ≃L[𝕜] E'` which belongs to the data of an immersion `f` at `x`:
+the particular equivalence is arbitrary, but this choice matches the witnesses given by
+`h.domChart` and `h.codChart`. -/
+noncomputable def equiv (h : IsImmersionAt F I I' n f x) : (E × F) ≃L[𝕜] E' :=
+  Classical.choose <| LiftSourceTargetPropertyAt.property h
+
+lemma writtenInCharts (h : IsImmersionAt F I I' n f x) :
+    EqOn ((h.codChart.extend I') ∘ f ∘ (h.domChart.extend I).symm) (h.equiv ∘ (·, 0))
+      (h.domChart.extend I).target :=
+  Classical.choose_spec <| LiftSourceTargetPropertyAt.property h
+
+lemma property (h : IsImmersionAt F I I' n f x) :
+    LiftSourceTargetPropertyAt I I' n f x (ImmersionAtProp F I I' M M') :=
+  IsImmersionAt.def.1 h
+
+/--
+Roig and Domingues' [roigdomingues1992] definition of immersions only asks for this inclusion
```

Comment:
Ideally, a docstring should be self-contained, i.e., readable without looking at the lemma statement. Here it's not the case, because "this inclusion" is not explained.